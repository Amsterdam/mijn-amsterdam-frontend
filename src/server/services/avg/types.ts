import { ZaakAanvraagDetail } from '../../../universal/types/App.types';
import { SmileSourceResponse, SmileFieldValue } from '../smile/smile-types';

export type AVGResponse = {
  aantal: number;
  verzoeken: AVGRequestFrontend[];
};

export interface AVGRequestFrontend extends ZaakAanvraagDetail {
  identifier: string;
  registratieDatum: string;
  type: string;
  toelichting: string;
  resultaat: string;
  ontvangstDatum: string;
  ontvangstDatumFormatted: string;
  opschortenGestartOp: string;
  datumInBehandeling: string;
  datumAfhandeling: string;
  themas: string;
}

export type SmileAvgResponse = SmileSourceResponse<SmileAVGRequest>;

export type SmileAVGRequest = {
  avgverzoek_id: SmileFieldValue;
  avgverzoek_statusavgverzoek_alias: SmileFieldValue;
  avgverzoek_datumbinnenkomst: SmileFieldValue;
  avgverzoek_typeverzoek: SmileFieldValue;
  avgverzoek_typeafhandelingvaststellen_resultaat: SmileFieldValue | null;
  avgverzoek_omschrijvingvanonderwerp: SmileFieldValue;
  avgverzoek_opschortengestartop: SmileFieldValue | null;
  avgverzoek_datuminbehandeling: SmileFieldValue | null;
  avgverzoek_werkelijkeeinddatum: SmileFieldValue | null;
};

export type SmileAvgThemesResponse = SmileSourceResponse<SmileAvgThemes>;

export type SmileAvgThemes = {
  themaperavgverzoek_avgthema_omschrijving: SmileFieldValue;
  themaperavgverzoek_avgverzoek_id: SmileFieldValue;
};

export type AvgTheme = {
  avgVerzoekId: string | null;
  themaOmschrijving: string | null;
};

export type AvgThemesResponse = {
  verzoeken: AvgTheme[];
};
