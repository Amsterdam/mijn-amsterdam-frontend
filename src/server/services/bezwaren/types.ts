import { GenericDocument, ZaakDetail } from '../../../universal/types';

export type kenmerkKey =
  | 'statustekst'
  | 'statusdatum'
  | 'resultaattekst'
  | 'besluitnr'
  | 'besluitdatum'
  | 'datumintrekking';

export type Kenmerk = { kenmerk: kenmerkKey; bron: string };

export interface BezwaarSourceData {
  uuid: string;
  // Ontvangstdatum
  registratiedatum: string;

  // Bezwaarnummer
  identificatie: string;

  // Onderwerp
  omschrijving: string | null;

  // Specificatie
  toelichting: string | null;

  // Status tekst
  status: string | null;

  // Datum van het resultaat van het bezwaar
  publicatiedatum: string | null;

  // In behandeling
  startdatum: string;

  einddatum: string | null;

  kenmerken: Kenmerk[];
}

export type Bezwaar = {
  identificatie: string;
  uuid: string;
  fetchUrl: string | null;
  startdatum: string;
  ontvangstdatum: string;
  ontvangstdatumFormatted: string | null;
  omschrijving: string | null;
  toelichting: string | null;
  statusdatum: string | null;
  datumResultaat: string | null;
  datumResultaatFormatted: string | null;
  einddatum: string | null;
  primairbesluit: string | null;
  primairbesluitdatum: string | null;
  primairbesluitdatumFormatted: string | null;
  resultaat: string | null;
  steps: never[];
  documenten: BezwaarDocument[];
} & ZaakDetail;

export type BezwaarSourceStatus = {
  url: string;
  uuid: string;
  zaak: string;
  statustype: string;
  datumStatusGezet: string;
  statustoelichting: string;
};

export interface BezwarenSourceResponse<T> {
  count: number;
  next: string;
  previous: string;
  results: T[];
}

export interface BezwaarSourceDocument {
  url: string;
  identificatie: string;
  bronorganisatie: string | null;
  creatiedatum: string;
  titel: string;
  vertrouwelijkheidaanduiding: string;
  auteur: string;
  status: string;
  formaat: string;
  taal: string;
  versie: number;
  beginRegistratie: string;
  bestandsnaam: string;
  inhoud: string;
  bestandsomvang: string | null;
  link: string | null;
  beschrijving: string;
  ontvangstdatum: string;
  verzenddatum: string;
  indicatieGebruiksrecht: string | null;
  verschijningsvorm: string | null;
  ondertekening: string | null;
  integriteit: string | null;
  informatieobjecttype: string;
  locked: boolean;
  dossiertype:
    | 'Online Correspondentie'
    | 'Online Procesdossier'
    | 'Online Besluitvorming'
    | 'Online Aangeleverd';
}

export type BezwaarDocument = GenericDocument & {
  dossiertype: BezwaarSourceDocument['dossiertype'];
};

export interface OctopusApiResponse<T> {
  items: T[];
  count: number;
}
