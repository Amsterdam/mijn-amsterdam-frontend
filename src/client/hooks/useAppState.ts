import { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { SetterOrUpdater, atom, useRecoilState, useRecoilValue } from 'recoil';
import { streamEndpointQueryParamKeys } from '../../universal/config/app';
import { FeatureToggle } from '../../universal/config/feature-toggles';
import {
  ApiPristineResponse,
  ApiResponse,
  ApiSuccessResponse,
  apiPristineResult,
} from '../../universal/helpers/api';
import {
  AppState,
  AppStateKey,
  BagThema,
} from '../../universal/types/App.types';
import { PRISTINE_APPSTATE, createAllErrorState } from '../AppState';
import { BFFApiUrls } from '../config/api';
import { transformSourceData } from '../data-transform/appState';
import { captureMessage } from '../utils/monitoring';
import { useDataApi } from './api/useDataApi';
import { useProfileTypeValue } from './useProfileType';
import { SSE_ERROR_MESSAGE, useSSE } from './useSSE';

const fallbackServiceRequestOptions = {
  postpone: true,
};

export const appStateAtom = atom<AppState>({
  key: 'appState',
  default: PRISTINE_APPSTATE as AppState,
});

interface useAppStateFallbackServiceProps {
  profileType: ProfileType;
  isEnabled: boolean;
  setAppState: SetterOrUpdater<AppState>;
}

export function useAppStateFallbackService({
  profileType,
  isEnabled,
  setAppState,
}: useAppStateFallbackServiceProps) {
  const [api, fetchFallbackService] = useDataApi<AppState | null>(
    fallbackServiceRequestOptions,
    null
  );

  const appStateError = useCallback(
    (message: string) => {
      captureMessage('Could not load any data sources.', {
        properties: {
          message,
        },
        severity: 'critical',
      });
      setAppState((appState) => createAllErrorState(appState, message));
    },
    [setAppState]
  );

  // If no EvenSource support or EventSource fails, the Fallback service endpoint is used for fetching all the data.
  useEffect(() => {
    if (!isEnabled) {
      return;
    }
    fetchFallbackService({
      ...fallbackServiceRequestOptions,
      url: BFFApiUrls.SERVICES_SAURON,
      postpone: false,
    });
  }, [fetchFallbackService, isEnabled, profileType]);

  // Update the appState with data fetched by the Fallback service endpoint
  useEffect(() => {
    if (!isEnabled) {
      return;
    }
    if (api.data !== null && !api.isLoading && !api.isError) {
      setAppState((appState) => {
        return Object.assign({}, appState, transformSourceData(api.data));
      });
    } else if (api.isError) {
      // If everything fails, this is the final state update.
      const errorMessage =
        'Services.all endpoint could not be reached or returns an error.';
      appStateError(errorMessage);
    }
  }, [api, appStateError, setAppState, isEnabled]);
}

export function addParamsToStreamEndpoint(
  url: string,
  searchParams: string = location.search
) {
  let streamEndpointUrl = url;

  // For testing and development purposes we can pass a set of arbitrary parameters to the BFF.
  // See also: universal/config/app.ts : streamEndpointQueryParamKeys
  if (FeatureToggle.passQueryParamsToStreamUrl) {
    let testStreamEndpointUrl = streamEndpointUrl;
    if (searchParams !== '') {
      const locationParams = new URLSearchParams(searchParams);
      const newUrlSearchParams = new URLSearchParams();
      for (const [param, value] of locationParams.entries()) {
        if (param in streamEndpointQueryParamKeys) {
          newUrlSearchParams.set(param, value);
        }
      }
      if (newUrlSearchParams.size) {
        streamEndpointUrl = `${testStreamEndpointUrl}?${newUrlSearchParams.toString()}`;
      }
    }
  }

  return streamEndpointUrl;
}

const streamEndpoint = addParamsToStreamEndpoint(BFFApiUrls.SERVICES_SSE);

/**
 * The primary communication is the EventSource. In the case the EventSource can't connect to the server, a number of retries will take place.
 * If the EventSource fails the Fallback service endpoint /all will be used in a last attempt to fetch the data needed to display a fruity application.
 * If that fails we just can't connect to the server for whatever reason. Monitoring error handling might have information about this scenario.
 */
export function useAppStateRemote() {
  const hasEventSourceSupport = 'EventSource' in window; // IE11 and early edge versions don't have EventSource support. These browsers will use the the Fallback service endpoint.
  const [isFallbackServiceEnabled, setFallbackServiceEnabled] = useState(
    !hasEventSourceSupport
  );

  const profileType = useProfileTypeValue();
  const [appState, setAppState] = useRecoilState(appStateAtom);

  // First retrieve all the services specified in the BFF, after that Only retrieve incremental updates
  const useIncremental = useRef(false);

  useEffect(() => {
    useIncremental.current = true;
  }, []);

  // The callback is fired on every incoming message from the EventSource.
  const onEvent = useCallback((messageData: any) => {
    if (messageData && messageData !== SSE_ERROR_MESSAGE) {
      const transformedMessageData = transformSourceData(messageData);
      setAppState((appState) => {
        const appStateUpdated = {
          ...appState,
          ...transformedMessageData,
        };
        return appStateUpdated;
      });
    } else if (messageData === SSE_ERROR_MESSAGE) {
      setFallbackServiceEnabled(true);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useSSE({
    path: streamEndpoint,
    eventName: 'message',
    callback: onEvent,
    postpone: isFallbackServiceEnabled,
  });

  useAppStateFallbackService({
    profileType,
    isEnabled: hasEventSourceSupport ? isFallbackServiceEnabled : true,
    setAppState,
  });

  return appState;
}

export function useAppStateGetter() {
  return useRecoilValue(appStateAtom);
}

export function useAppStateSetter() {
  return useRecoilState(appStateAtom)[1];
}

export function isAppStateReady(
  appState: AppState,
  pristineAppState: AppState,
  profileType: ProfileType
) {
  const isLegacyProfileType = ['private', 'commercial'].includes(profileType);
  const profileStates = Object.entries(appState).filter(
    ([appStateKey, state]) => {
      const key = appStateKey as AppStateKey;
      const stateConfig = pristineAppState[
        key
      ] as unknown as ApiPristineResponse<any>;

      const isProfileMatch =
        (isLegacyProfileType && !stateConfig?.profileTypes?.length) ||
        stateConfig?.profileTypes?.includes(profileType);

      // NOTE: The appState keys ending with _BAG are not considered a fixed/known portion of the appstate.
      if (!stateConfig && !key.endsWith('_BAG')) {
        captureMessage(`unknown stateConfig key: ${appStateKey}`);
      }

      // If we encounter an unknown stateConfig we treat the state to be ready so we don't block the isReady completely.
      return isProfileMatch && (!!stateConfig?.isActive || !stateConfig);
    }
  );

  return (
    !!profileStates.length &&
    profileStates.every(([appStateKey, state]) => {
      return typeof state !== 'undefined' && state.status !== 'PRISTINE';
    })
  );
}

export function useAppStateReady() {
  const appState = useAppStateGetter();
  const profileType = useProfileTypeValue();

  return useMemo(
    () => isAppStateReady(appState, PRISTINE_APPSTATE, profileType),
    [appState, profileType]
  );
}

export interface AppStateBagApiParams {
  url: string;
  bagThema: BagThema;
  key: string;
  postponeFetch?: boolean;
}

// Use this hook for loading additional data that needs to be persisted in the state. For example additional data loaded if a user navigates to a detailpage
// that requires fetching data that wasn't retrieved initially.
export function useAppStateBagApi<T extends unknown>({
  url,
  bagThema,
  key,
  postponeFetch = false,
}: AppStateBagApiParams) {
  const [appState, setAppState] = useRecoilState(appStateAtom);
  const isApiDataCached =
    typeof appState[bagThema] !== null &&
    typeof appState[bagThema] === 'object' &&
    typeof appState[bagThema] !== 'undefined' &&
    key in appState[bagThema]!;

  const [api, fetch] = useDataApi<ApiResponse<T | null>>(
    {
      url,
      postpone: isApiDataCached || !url,
    },
    apiPristineResult(null)
  );

  useEffect(() => {
    if (url && !isApiDataCached && !api.isDirty && !api.isLoading) {
      fetch({
        url,
        postpone: false,
      });
    }
    if (!isApiDataCached && !!api.data.content) {
      setAppState((state) => {
        let localState = state[bagThema];
        if (!localState) {
          localState = {};
        }

        localState = {
          ...localState,
          [key]: api.data.content as T,
        };

        if ((api.data as ApiSuccessResponse<T>).failedDependencies) {
          localState = {
            ...localState,
            failedDependencies: (api.data as ApiSuccessResponse<T>)
              .failedDependencies,
          };
        }

        return {
          ...state,
          [bagThema]: localState,
        };
      });
    }
  }, [isApiDataCached, api, key, url]);

  return [appState?.[bagThema]?.[key] as T, api, fetch] as const;
}

export function useRemoveAppStateBagData() {
  const [appState, setAppState] = useRecoilState(appStateAtom);
  return useCallback(
    ({ bagThema, key: keyExpected }: Omit<AppStateBagApiParams, 'url'>) => {
      const local = appState[bagThema];
      if (!!local) {
        setAppState(
          Object.assign({}, appState, {
            [bagThema]: Object.fromEntries(
              Object.entries(local).filter(([key]) => {
                return keyExpected !== key;
              })
            ),
          })
        );
      }
    },
    [appState]
  );
}
