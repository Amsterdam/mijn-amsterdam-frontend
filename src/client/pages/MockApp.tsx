import { RecoilRoot, RecoilRootProps } from 'recoil';
import React from 'react';
import { MemoryRouter, Route } from 'react-router-dom';

interface MockAppProps {
  routePath: string;
  routeEntry: string;
  initializeState?: RecoilRootProps['initializeState'];
  component: React.Component | any;
}

export default function MockApp({
  routePath,
  routeEntry,
  initializeState,
  component,
}: MockAppProps) {
  return (
    <RecoilRoot initializeState={initializeState}>
      <MemoryRouter initialEntries={[routeEntry]}>
        <Route path={routePath} component={component} />
      </MemoryRouter>
    </RecoilRoot>
  );
}
