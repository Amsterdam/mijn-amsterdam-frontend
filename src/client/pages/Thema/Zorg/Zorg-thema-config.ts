import { generatePath } from 'react-router';

import { WMOVoorzieningFrontend } from '../../../../server/services/wmo/wmo-config-and-types';
import { LinkProps } from '../../../../universal/types/App.types';
import { withOmitDisplayPropsForSmallScreens } from '../../../components/Table/helpers';
import {
  DisplayProps,
  WithDetailLinkComponent,
} from '../../../components/Table/TableV2.types';
import { MAX_TABLE_ROWS_ON_THEMA_PAGINA } from '../../../config/app';
import type { ThemaRoutesConfig } from '../../../config/thema-types';

const MAX_TABLE_ROWS_ON_THEMA_PAGINA_HUIDIG = 5;
const MAX_TABLE_ROWS_ON_THEMA_PAGINA_EERDER = MAX_TABLE_ROWS_ON_THEMA_PAGINA;

export const listPageParamKind = {
  actual: 'huidige-voorzieningen',
  historic: 'eerdere-en-afgewezen-voorzieningen',
} as const;

type ListPageParamKey = keyof typeof listPageParamKind;
export type ListPageParamKind = (typeof listPageParamKind)[ListPageParamKey];

export const featureToggle = {
  zorgActive: true,
};

export const themaId = 'ZORG' as const;
export const themaTitle = 'Zorg en ondersteuning';

export const routeConfig = {
  detailPage: {
    path: '/zorg-en-ondersteuning/voorziening/:id',
    trackingUrl: '/zorg-en-ondersteuning/voorziening',
    documentTitle: `Voorziening | ${themaTitle}`,
  },
  listPage: {
    path: '/zorg-en-ondersteuning/lijst/:kind/:page?',
    documentTitle: (params) =>
      `${params?.kind === listPageParamKind.actual ? 'Huidige' : 'Eerdere en afgewezen'} voorzieningen | ${themaTitle}`,
  },
  themaPage: {
    path: '/zorg-en-ondersteuning',
    documentTitle: `${themaTitle} | overzicht`,
  },
} as const satisfies ThemaRoutesConfig;

export const linkListItems: LinkProps[] = [
  {
    to: 'https://www.amsterdam.nl/zorg-ondersteuning/',
    title: 'Lees hier meer over zorg en ondersteuning',
  },
];

const displayPropsBase: DisplayProps<
  WithDetailLinkComponent<WMOVoorzieningFrontend>
> = {
  detailLinkComponent: 'Naam',
  displayStatus: 'Status',
  statusDateFormatted: 'Datum',
};

const displayProps = withOmitDisplayPropsForSmallScreens(displayPropsBase, [
  'displayStatus',
  'statusDateFormatted',
]);

export const listPageTitle = {
  [listPageParamKind.actual]: 'Huidige voorzieningen',
  [listPageParamKind.historic]: 'Eerdere en afgewezen voorzieningen',
} as const;

export const tableConfig = {
  [listPageParamKind.actual]: {
    title: listPageTitle[listPageParamKind.actual],
    filter: (regeling: WMOVoorzieningFrontend) => regeling.isActual,
    displayProps: displayProps,
    maxItems: MAX_TABLE_ROWS_ON_THEMA_PAGINA_HUIDIG,
    textNoContent: 'U heeft geen huidige voorzieningen.',
    listPageRoute: generatePath(routeConfig.listPage.path, {
      kind: listPageParamKind.actual,
      page: null,
    }),
  },
  [listPageParamKind.historic]: {
    title: listPageTitle[listPageParamKind.historic],
    filter: (regeling: WMOVoorzieningFrontend) => !regeling.isActual,
    displayProps: displayProps,
    maxItems: MAX_TABLE_ROWS_ON_THEMA_PAGINA_EERDER,
    textNoContent:
      'U heeft geen eerdere en/of afgewezen voorzieningen. U ziet hier niet alle gegevens uit het verleden. De gegevens die u hier niet ziet, heeft u eerder per post ontvangen.',
    listPageRoute: generatePath(routeConfig.listPage.path, {
      kind: listPageParamKind.historic,
      page: null,
    }),
  },
} as const;
