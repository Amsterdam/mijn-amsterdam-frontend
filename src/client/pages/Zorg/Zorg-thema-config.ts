import { WMOVoorzieningFrontend } from '../../../server/services/wmo/wmo-config-and-types';
import { AppRoutes } from '../../../universal/config/routes';
import { dateSort } from '../../../universal/helpers/date';
import { MAX_TABLE_ROWS_ON_THEMA_PAGINA } from '../../config/app';
import styles from './Zorg.module.scss';

const MAX_TABLE_ROWS_ON_THEMA_PAGINA_HUIDIG = 5;
const MAX_TABLE_ROWS_ON_THEMA_PAGINA_EERDER = MAX_TABLE_ROWS_ON_THEMA_PAGINA;

const displayProps = {
  detailLinkComponent: '',
  status: 'Status',
  dateDescisionFormatted: 'Datum',
};

export const listPageParamKind = {
  actual: 'huidige-regelingen',
  historic: 'eerdere-en-afgewezen-regelingen',
} as const;

export type ListPageParamKey = keyof typeof listPageParamKind;
export type ListPageParamKind = (typeof listPageParamKind)[ListPageParamKey];

export const listPageTitle = {
  [listPageParamKind.actual]: 'Huidige regelingen',
  [listPageParamKind.historic]: 'Eerdere en afgewezen regelingen',
} as const;

export const tableConfig = {
  [listPageParamKind.actual]: {
    title: listPageTitle[listPageParamKind.actual],
    filter: (regeling: WMOVoorzieningFrontend) => regeling.isActual,
    sort: dateSort('dateDecision', 'desc'),
    displayProps: displayProps,
    maxItems: MAX_TABLE_ROWS_ON_THEMA_PAGINA_HUIDIG,
    className: styles.HuidigeRegelingen,
  },
  [listPageParamKind.historic]: {
    title: listPageTitle[listPageParamKind.historic],
    filter: (regeling: WMOVoorzieningFrontend) => !regeling.isActual,
    sort: dateSort('dateDecision', 'desc'),
    displayProps: displayProps,
    maxItems: MAX_TABLE_ROWS_ON_THEMA_PAGINA_EERDER,
    className: styles.EerdereRegelingen,
  },
} as const;

export const routes = {
  listPage: AppRoutes['ZORG/VOORZIENINGEN_LIST'],
  detailPage: AppRoutes['ZORG/VOORZIENING'],
  themaPage: AppRoutes['ZORG'],
} as const;
