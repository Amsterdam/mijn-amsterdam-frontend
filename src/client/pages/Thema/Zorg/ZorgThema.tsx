import { Link, Paragraph } from '@amsterdam/design-system-react';

import { useZorgThemaData } from './useZorgThemaData';
import { WMOVoorzieningFrontend } from '../../../../server/services/wmo/wmo-config-and-types';
import { PageContentCell } from '../../../components/Page/Page';
import { ParagaphSuppressed } from '../../../components/ParagraphSuppressed/ParagraphSuppressed';
import ThemaPagina from '../../../components/Thema/ThemaPagina';
import ThemaPaginaTable from '../../../components/Thema/ThemaPaginaTable';
import { useHTMLDocumentTitle } from '../../../hooks/useHTMLDocumentTitle';

export const WMO_HELPDESK_PHONENUMBER = '0800 0643' as const;
export const WMO_HELPDESK_HREF_TEL_LINK =
  `tel:${WMO_HELPDESK_PHONENUMBER.replace(' ', '')}` as const;

export function HistoricItemsMention() {
  return (
    <ParagaphSuppressed>
      U ziet hier informatie vanaf 1 januari 2018. Wilt u iets weten van
      daarvoor? Bel dan de Wmo Helpdesk:{' '}
      <Link href={WMO_HELPDESK_HREF_TEL_LINK} rel="noreferrer">
        {WMO_HELPDESK_PHONENUMBER}
      </Link>
      . U ziet ook geen begeleid thuis en beschermd verblijf (vroeger
      maatschappelijke opvang en beschermd wonen). Hier zijn we nog mee bezig.
    </ParagaphSuppressed>
  );
}

export function ZorgThema() {
  const {
    isError,
    isLoading,
    voorzieningen,
    id,
    title,
    tableConfig,
    linkListItems,
    routeConfig,
  } = useZorgThemaData();
  useHTMLDocumentTitle(routeConfig.themaPage);

  const pageContentTop = (
    <PageContentCell spanWide={8}>
      <Paragraph className="ams-mb-m">
        Hieronder ziet u uw voorzieningen vanuit de Wet maatschappelijke
        ondersteuning (Wmo).
      </Paragraph>
      <Paragraph>
        Heeft u vragen of wilt u een wijziging doorgeven? <br />
        Bel dan gratis de Wmo Helpdesk:{' '}
        <Link rel="noreferrer" href={WMO_HELPDESK_HREF_TEL_LINK}>
          {WMO_HELPDESK_PHONENUMBER}
        </Link>{' '}
        (maandag tot en met vrijdag van 08.00 tot 18.00 uur) of ga langs bij het
        Sociaal Loket.
      </Paragraph>
    </PageContentCell>
  );

  const tables = Object.entries(tableConfig).map(
    ([
      kind,
      { title, displayProps, textNoContent, filter, maxItems, listPageRoute },
    ]) => {
      return (
        <ThemaPaginaTable<WMOVoorzieningFrontend>
          key={kind}
          title={title}
          zaken={voorzieningen.filter(filter)}
          listPageRoute={listPageRoute}
          displayProps={displayProps}
          textNoContent={textNoContent}
          maxItems={maxItems}
        />
      );
    }
  );

  return (
    <>
      <ThemaPagina
        id={id}
        title={title}
        pageContentTop={pageContentTop}
        linkListItems={linkListItems}
        pageContentMain={
          <>
            {tables}
            <PageContentCell spanWide={8} startWide={3}>
              <HistoricItemsMention />
            </PageContentCell>
          </>
        }
        isError={isError}
        isLoading={isLoading}
      />
    </>
  );
}
