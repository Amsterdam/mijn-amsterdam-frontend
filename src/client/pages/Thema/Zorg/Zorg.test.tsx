import { render } from '@testing-library/react';
import { generatePath } from 'react-router';
import { MutableSnapshot } from 'recoil';

import { routeConfig } from './Zorg-thema-config.ts';
import { ZorgThema } from './ZorgThema.tsx';
import { AppState } from '../../../../universal/types/App.types.ts';
import { appStateAtom } from '../../../hooks/useAppState.ts';
import MockApp from '../../MockApp.tsx';

const testState: Pick<AppState, 'WMO'> = {
  WMO: {
    status: 'OK',
    content: [
      {
        id: 'wmo-item-1',
        title: 'Wmo item 1',
        supplier: 'Mantelzorg B.V',
        dateDecision: '2020-07-24',
        dateDecisionFormatted: '24 juli 2020',
        statusDate: '2020-09-24',
        statusDateFormatted: '24 september 2020',
        isActual: true,
        link: {
          to: 'http://example.org/ding',
          title: 'Linkje!! naar wmo item 1',
        },
        displayStatus: 'Levering gestart',
        decision: 'Klaar',
        documents: [],
        itemTypeCode: 'WMO',
        steps: [
          {
            id: 'wmo-step-2',
            status: 'Besluit',
            datePublished: '2020-08-08',
            description: 'De levering van uw thuiszorg is gestart',
            documents: [],
            isActive: false,
            isChecked: true,
          },
          {
            id: 'wmo-step-1',
            status: 'Levering gestart',
            datePublished: '2020-09-24',
            description: 'De levering van uw thuiszorg is gestart',
            documents: [],
            isActive: true,
            isChecked: true,
          },
        ],
      },
    ],
  },
};

function initializeState(snapshot: MutableSnapshot) {
  snapshot.set(appStateAtom, testState as AppState);
}

describe('<Zorg />', () => {
  const routeEntry = generatePath(routeConfig.themaPage.path);
  const routePath = routeConfig.themaPage.path;

  function Component() {
    return (
      <MockApp
        routeEntry={routeEntry}
        routePath={routePath}
        component={ZorgThema}
        initializeState={initializeState}
      />
    );
  }

  it('Matches the Full Page snapshot', () => {
    const { asFragment } = render(<Component />);
    expect(asFragment()).toMatchSnapshot();
  });
});
