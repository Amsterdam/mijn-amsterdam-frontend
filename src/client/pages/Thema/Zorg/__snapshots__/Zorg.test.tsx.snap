// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Zorg /> > Matches the Full Page snapshot 1`] = `
<DocumentFragment>
  <div
    class="ams-grid"
  >
    <div
      class="ams-grid__cell ams-grid__cell--span-4 ams-grid__cell--span-8-medium ams-grid__cell--span-12-wide ams-grid__cell--start-1 ams-grid__cell--start-1-medium ams-grid__cell--start-1-wide _PageHeadingWrap_c5a3f1"
    >
      <div
        class="_PageHeadingInner_c5a3f1"
        id="skip-to-id-AppContent"
      >
        <nav
          class="ams-breadcrumb"
        >
          <ol
            class="ams-breadcrumb__list"
          >
            <li
              class="ams-breadcrumb__item"
            >
              <a
                class="ams-breadcrumb__link"
                href="/"
              >
                Home
              </a>
            </li>
            <li
              class="ams-breadcrumb__item"
            >
              <a
                class="ams-breadcrumb__link"
                href="/zorg-en-ondersteuning"
                tabindex="-1"
              >
                <span
                  class="ams-visually-hidden"
                >
                  Zorg en ondersteuning
                </span>
              </a>
            </li>
          </ol>
        </nav>
        <h3
          class="ams-heading ams-heading--1 _PageHeading_c5a3f1"
        >
          Zorg en ondersteuning
        </h3>
      </div>
    </div>
    <div
      class="ams-grid__cell ams-grid__cell--span-4 ams-grid__cell--span-8-medium ams-grid__cell--span-8-wide ams-grid__cell--start-1 ams-grid__cell--start-1-medium ams-grid__cell--start-1-wide"
    >
      <p
        class="ams-paragraph ams-mb-m"
      >
        Hieronder ziet u uw voorzieningen vanuit de Wet maatschappelijke ondersteuning (Wmo).
      </p>
      <p
        class="ams-paragraph"
      >
        Heeft u vragen of wilt u een wijziging doorgeven? 
        <br />
        Bel dan gratis de Wmo Helpdesk: 
        <a
          class="ams-link"
          href="tel:08000643"
          rel="noreferrer"
        >
          0800 0643
        </a>
         (maandag tot en met vrijdag van 08.00 tot 18.00 uur) of ga langs bij het Sociaal Loket.
      </p>
    </div>
    <div
      class="ams-grid__cell ams-grid__cell--span-4 ams-grid__cell--span-8-medium ams-grid__cell--span-12-wide ams-grid__cell--start-1 ams-grid__cell--start-1-medium ams-grid__cell--start-1-wide"
    >
      <ul
        class="ams-link-list"
      >
        <li>
          <a
            class="ams-link-list__link"
            href="https://www.amsterdam.nl/zorg-ondersteuning/"
            rel="noreferrer"
          >
            <span
              class="ams-icon"
            >
              <svg
                aria-hidden="true"
                focusable="false"
                viewBox="0 0 32 32"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m9.757 32-2.9-2.91L19.937 16 6.857 2.91 9.757 0l16 16z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
            Lees hier meer over zorg en ondersteuning
          </a>
        </li>
      </ul>
    </div>
    <div
      class="ams-grid__cell ams-grid__cell--span-4 ams-grid__cell--span-8-medium ams-grid__cell--span-12-wide ams-grid__cell--start-1 ams-grid__cell--start-1-medium ams-grid__cell--start-1-wide"
    >
      <h3
        class="ams-heading ams-heading--2"
      >
        Huidige voorzieningen
      </h3>
      <div
        class="ams-table"
      >
        <table
          class="ams-table__table _TableV2_4c6825"
        >
          <thead
            class="ams-table__header"
          >
            <tr
              class="ams-table__row"
            >
              <th
                class="ams-table__header-cell"
              >
                Naam
              </th>
              <th
                class="ams-table__header-cell"
              >
                Status
              </th>
              <th
                class="ams-table__header-cell"
              >
                Datum
              </th>
            </tr>
          </thead>
          <tbody
            class="ams-table__body"
          >
            <tr
              class="ams-table__row"
            >
              <td
                class="ams-table__cell"
              >
                <a
                  class="ams-link _MaLink_aa758a _MaRouterLink__no-underline-fat_aa758a"
                  href="http://example.org/ding"
                  title="Bekijk meer over Wmo item 1"
                >
                  Wmo item 1
                </a>
              </td>
              <td
                class="ams-table__cell"
              >
                Levering gestart
              </td>
              <td
                class="ams-table__cell"
              >
                24 september 2020
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      class="ams-grid__cell ams-grid__cell--span-4 ams-grid__cell--span-8-medium ams-grid__cell--span-12-wide ams-grid__cell--start-1 ams-grid__cell--start-1-medium ams-grid__cell--start-1-wide"
    >
      <h3
        class="ams-heading ams-heading--2"
      >
        Eerdere en afgewezen voorzieningen
      </h3>
      <div
        class="ams-table"
      >
        <table
          class="ams-table__table _TableV2_4c6825"
        >
          <tbody
            class="ams-table__body"
          />
        </table>
      </div>
      <p
        class="ams-paragraph"
      >
        U heeft geen eerdere en/of afgewezen voorzieningen. U ziet hier niet alle gegevens uit het verleden. De gegevens die u hier niet ziet, heeft u eerder per post ontvangen.
      </p>
    </div>
    <div
      class="ams-grid__cell ams-grid__cell--span-4 ams-grid__cell--span-8-medium ams-grid__cell--span-8-wide ams-grid__cell--start-1 ams-grid__cell--start-1-medium ams-grid__cell--start-3-wide"
    >
      <p
        class="ams-paragraph _SuppressedParagraph_2fb90c"
      >
        U ziet hier informatie vanaf 1 januari 2018. Wilt u iets weten van daarvoor? Bel dan de Wmo Helpdesk: 
        <a
          class="ams-link"
          href="tel:08000643"
          rel="noreferrer"
        >
          0800 0643
        </a>
        . U ziet ook geen begeleid thuis en beschermd verblijf (vroeger maatschappelijke opvang en beschermd wonen). Hier zijn we nog mee bezig.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
